Add user function definition support

lenticularis39 opened the pull request:
A new structure for defining subprograms is added on the same level as probes:
```
fn <function_name>(<type0> <arg0>, ...): <return_type> {
  <body>
}
```
The existing return statement is extended to also allow returning a specified value. SemanticAnalyser checks if the type is correct, while CodegenLLVM goes through the generated code to ensure all code flow paths return a value in a non-void function.

Only a limited number of builtins are allowed to be used inside functions due to current implementation restrictions; these can be addressed in later development.

<!--
Please provide a description of your change below this comment.

Then please complete the checklist.

Useful contribution guidelines and tips are in docs/developers.md.
-->

----

This is the first step in implementing #2516 that can be either merged separately or together with the rest of the implementation. Note that this does not implement a way to call the functions nor a way to generate the correct BPF code for that.

(I'm opening this draft PR, since I believe there is quite a few non-trivial changes that it makes sense to review while I work on the rest.)

##### Checklist

Note: this checklist is only for the function definition part, excluding the BPF code generator and function calls.

- [x] Tests for parser
- [x] Tests for semantic analyser
- [x] Tests for codegen
- [x] Function argument support

viktormalik commented on tests/parser.cpp:
```diff
@@ -1971,6 +1971,121 @@ TEST(Parser, invalid_provider)
+{
+  test("fn f1(): void {}",
+       "Program\n"
+       " f1: void\n");
```
Nit: would it be possible to add `()` here so that it is visible that it's a function?
```suggestion
       " f1: void()\n");
```

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2465,6 +2472,72 @@ void CodegenLLVM::generateProbe(Probe &probe,
+
+  if (!subprog.type.IsVoidTy())
+  {
+    // Check that every reachable return is non-dummy (for dummy returns, see
```
This piece of code would deserve to be enclosed in a separate function, e.g. `checkMissingReturns` or so.

viktormalik reviewed (commented):
Looks great, the implementation is pretty straightforward. Just a couple of comments.

I'd advocate for merging this before implementing support for function calls.

lenticularis39 commented:
Resolved comments and rebased.

viktormalik commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2465,6 +2472,72 @@ void CodegenLLVM::generateProbe(Probe &probe,
+        bfs_visited.insert(succ);
+        bfs_queue.push_front(succ);
+      }
+    }
```
This still applies, right?

lenticularis39 replied:
Ah yes, I somehow missed that you are suggesting to run unreachable code elimination. The codegen already generates dummy returns for probes, they are just simpler, since the return value of probes is either void or some kind of integer, so there is no need for an empty alloca.

On the other hand, running dead code elimination here might slightly impact the tests (e.g. the pass might fail on invalid LLVM IR or errors in unreachable code won't be caught in verification). I will try to put the pass there and see what it does.

lenticularis39 replied:
I tried running `UnreachableBlockElimPass` at the end of `void CodegenLLVM::visit(Subprog &subprog)` and it looks good. I think it might make sense to do that in a separate PR together with the same pass for probes rather than here, as I wrote above, this is not an issue specific to user-defined functions, it is just the most visible there.

ajor commented on src/ast/visitors.cpp:
I think Mutator would need to visit jump.return_value to be fully correct.
Doesn't it also need to visit subprog.args?

However, it seems like Mutator & Dispatcher aren't used anywhere, so this is all dead code anyway. This also means we can't test it.

I don't want to merge in potentially incorrect code, so I think we'd be better off deleting these classes entirely.

cc @fbs

lenticularis39 replied:
I'm not sure myself what the Mutator class is supposed to do exactly, you are probably right.

ajor replied:
After thinking about this some more and discussing the Mutator with @fbs, I think we're going to remove it entirely for now as it's not used. (it can be brought back when needed)

Basically: don't worry about the mutator in this PR

lenticularis39 replied:
OK, thanks.

fbs replied:
goal of the mutator is to write ast modifying passes, it mutates one ast into another. But never got around to writing mutating passes

ajor commented on src/lexer.l:
```diff
@@ -183,19 +191,41 @@ struct|union|enum       {
+  \n                    buffer += '\n'; loc.lines(1); loc.step();
+}
+<STRUCT,BRACE>{
+  "*"|")"|","|{space}+"$" {
```
Why are the space changes needed in lexer.l (this line and also in the `AFTER_COLON` section)? Why is space handled differently in this section?

lenticularis39 replied:
`AFTER_COLON` is an `%x` section, IIUC the usual rules do not apply there. Here the space handling should be a part of the logic ensuring `struct test $x` (for example) is parsed correctly.

I can take a closer look at it if that did not make much sense, that's what I remember from writing this code.

ajor reviewed (changes requested):
Love the tests!

The one main structural comment I've got is that something like checking all paths return a value should come under semantic analysis rather than codegen. Codegen should be able to assume that what it's been given is valid, and not worry about checking things itself.

Slightly off topic: We need to start caring more about the performance of the parser. Our approach previously was to not worry because it's not on the tracing hot-path, but the unit tests taking multiple seconds to run indicates that we've taken this too far. I did a quick profile and we're spending a lot of our time just copying strings around. I left a few comments on the new changes here, but we'll need to revisit the rest of ast.h at some point.

danobi commented:
Going to set aside some time this weekend to take a look

lenticularis39 commented on tests/parser.cpp:
```diff
@@ -1971,6 +1971,121 @@ TEST(Parser, invalid_provider)
   test_parse_failure("asdf:xyz:www { }");
 }
 
+TEST(Parser, subprog_only)
```
Oops, I missed that.

lenticularis39 commented on src/ast/ast.h:
```diff
@@ -541,6 +547,16 @@ class While : public Statement
+{
+public:
+  Scope(StatementList *stmts);
+  Scope(const Scope &other) = default;
```
Ah, that is true, and I know that - no idea why I put the line there, might have been copied from somewhere and changed because of compiler warnings. Thanks!

lenticularis39 commented on src/ast/ast.h:
```diff
@@ -541,6 +547,16 @@ class While : public Statement
+public:
+  Scope(StatementList *stmts);
+  Scope(const Scope &other) = default;
+  virtual ~Scope();
```
I agree, that was the intention - I was just focusing more on the functional side of things and forgot to change it.

lenticularis39 commented:
> Love the tests!

Thank you! I had to write thorough tests, the area of possibility of bugs here is almost unlimited :D

> 
> The one main structural comment I've got is that something like checking all paths return a value should come under semantic analysis rather than codegen. Codegen should be able to assume that what it's been given is valid, and not worry about checking things itself.
> 

I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the *AST*, while this checks for a property of the *CFG*. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.

> Slightly off topic: We need to start caring more about the performance of the parser. Our approach previously was to not worry because it's not on the tracing hot-path, but the unit tests taking multiple seconds to run indicates that we've taken this too far. I did a quick profile and we're spending a lot of our time just copying strings around. I left a few comments on the new changes here, but we'll need to revisit the rest of ast.h at some point.

Good point.

ajor commented:
> I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the _AST_, while this checks for a property of the _CFG_. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.

In languages with more complex control flow (e.g. `goto` statements) I think this makes sense, but is it necessary for bpftrace, which only has structured control flow?

I feel like something similar to this pseudo-code should be able to detect missing return statements:
```
fn all_paths_return(block):
  for stmt in block:
    if stmt == "return":
      return true
    if stmt == "break" or stmt == "continue":
      return false
    if stmt == "if":
      if all_paths_return(stmt.true_branch) && all_paths_return(stmt.false_branch):
        return true
    if stmt has children: (e.g. a loop)
      if all_paths_return(stmt.children):
        return true
  return false
```

danobi commented on src/ast/ast.h:
```diff
@@ -615,16 +631,60 @@ class Probe : public Node {
+  ~Subprog();
+
+  SubprogArgList *args = nullptr;
+  SizedType type;
```
nit: maybe `return_type` instead of `type`? To differentiate between type of function and return type

danobi commented on src/ast/passes/codegen_llvm.h:
```diff
@@ -230,6 +241,7 @@ class CodegenLLVM : public Visitor
   int next_probe_index_ = 1;
   // Used if there are duplicate USDT entries
   int current_usdt_location_index_{ 0 };
+  bool inside_subprog_;
```
nit: just to make it explicit
```suggestion
  bool inside_subprog_ = false;
```

danobi commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2537,6 +2603,33 @@ void CodegenLLVM::createRet(Value *value)
   }
 }
 
+void CodegenLLVM::checkMissingReturns(Function *func, Subprog &subprog)
+{
+  auto start_bb = &func->getEntryBlock();
+  std::deque<llvm::BasicBlock *> bfs_queue{ start_bb };
+  std::unordered_set<llvm::BasicBlock *> bfs_visited{ start_bb };
+  while (!bfs_queue.empty())
+  {
+    auto bb = bfs_queue.back();
+    bfs_queue.pop_back();
+    auto ret = llvm::dyn_cast<llvm::ReturnInst>(bb->getTerminator());
+    if (ret && ret->getMetadata("ret_dummy"))
+    {
+      LOG(ERROR, subprog.loc, err_) << "Not all paths in function "
+                                    << subprog.name() << " return a value";
+    }
+    auto term = bb->getTerminator();
+    for (unsigned i = 0, e = term->getNumSuccessors(); i != e; i++)
+    {
+      auto succ = term->getSuccessor(i);
+      if (bfs_visited.find(succ) != bfs_visited.end())
+        continue;
+      bfs_visited.insert(succ);
+      bfs_queue.push_front(succ);
+    }
+  }
+}
```
Same comment as @ajor -- doesn't feel right to do this in codegen

danobi commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2502,6 +2550,24 @@ void CodegenLLVM::createRet(Value *value)
+      b_.CreateRetVoid();
+    else
+    {
+      auto dummy_alloca = b_.CreateAlloca(ret_type, nullptr, "ret_dummy");
+      auto dummy = b_.CreateLoad(ret_type, dummy_alloca);
+      auto ret_inst = b_.CreateRet(dummy);
+      ret_inst->setMetadata("ret_dummy", MDTuple::get(b_.getContext(), {}));
```
Does moving return checks into semantic analyser get rid of the ret_dummy stuff? Not immediately clear to me what this does

danobi commented on tests/semantic_analyser.cpp:
```diff
@@ -2702,6 +2702,20 @@ TEST(semantic_analyser, string_size)
+
+TEST(semantic_analyser, subprog_arguments)
+{
+  test("fn f(int64 $a): int64 { return $a; }", 0);
```
Sorry if this was discussed elsewhere already, but the argument syntax feels a bit odd to me.

The function syntax is `fn <idents and stuff>: <ret type>`. It seems to me it would follow that argument syntax should also be `<ident>: <type>`, but here it's reversed.

IIRC we also added type annotations for local variables as `$var: type = <blah>` right?

danobi replied:
In that case I am in favor of `$arg: type` syntax

fbs replied:
I wonder if this type syntax can be expanded to other cases, like:

```
kprobe:some_kernel_fn(a: int, b: int): int
```

Basically adding the option to specify function signatures in places (althoug with btf and dwarf that might be a bit late). In that case, i wonder if having `->` like rust,python will make parsing easier

fbs commented:
> > I also thought that initially, however, checking if all paths return a value is actually not semantic analysis, but control flow analysis, since semantic analysis operates on the _AST_, while this checks for a property of the _CFG_. Implementing it in semantic analyser would require to construct the control flow inside the AST, which would duplicate work that is already done in codegen.
> 
> In languages with more complex control flow (e.g. `goto` statements) I think this makes sense, but is it necessary for bpftrace, which only has structured control flow?
> 
> I feel like something similar to this pseudo-code should be able to detect missing return statements:
> 
> ```
> fn all_paths_return(block):
>   for stmt in block:
>     if stmt == "return":
>       return true
>     if stmt == "break" or stmt == "continue":
>       return false
>     if stmt == "if":
>       if all_paths_return(stmt.true_branch) && all_paths_return(stmt.false_branch):
>         return true
>     if stmt has children: (e.g. a loop)
>       if all_paths_return(stmt.children):
>         return true
>   return false
> ```

I'd put something like this in its own pass. Is not semantic but also shouldn't go in the codegen. Passes should be fairly easy to add as all the boilerplate is handled (see e.g. nodecount.h).

Might be a good usecase for the `Dispatcher`  visitor as it allows for a return value which makes it a bit easier to write. 

Might be good to do it in a separte PR to avoid some noise in this one

lenticularis39 commented:
> I'd put something like this in its own pass. Is not semantic but also shouldn't go in the codegen. Passes should be fairly easy to add as all the boilerplate is handled (see e.g. nodecount.h).
> 
> Might be a good usecase for the `Dispatcher` visitor as it allows for a return value which makes it a bit easier to write.
> 
That sounds like the best idea to me, that will avoid the check being in codegen and having it returning bool, and also complicating the code of semantic analyzer.

Also, there is a direct benefit of doing it on the AST as  @ajor and @danobi suggested: we can print the exact place where a return instruction should be and isn't, which will make it easier for the programmer.

> Might be good to do it in a separte PR to avoid some noise in this one
There might be just an issue with the fact that without the check, bpftrace will generate invalid LLVM IR; however, I don't think that is that big of an issue, since the function feature is not officially supported yet. I can also make it a separate commit, that might be better.

fbs commented:
The idea behind the passmanager and all the visitor stuff is that it should make new passes easy to add, to avoid increasing the complexity of existing passes, the semantic analyser is quite tricky in places.

If you do want to try the dispather visitor, it will be the first usecase and its a bit hacky so it might not work that  well :(


Also, really nice work here :)

lenticularis39 commented:
Thank you all for the helpful comments, I feel like it cleared a lot of things up and will increase the quality of the PR a lot.

lenticularis39 commented:
Resolved (hopefully) all of the comments. There are some minor issues with the code like leftover declarartions, I will clean it after rebasing to current master.

danobi commented:
Top of my todo list to review

viktormalik commented on tests/return_path_analyser.cpp:
```diff
@@ -0,0 +1,100 @@
+
+TEST(return_path_analyser, if_else)
+{
+  test("fn test($x: int64): int64 { if ($x > 0) { return 0; } else { return 1; "
+       "} }",
```
Can we do something with the ugly linebreaks or is it clang-format enforcing this?
```suggestion
  test("fn test($x: int64): int64 {"
       "  if ($x > 0) { return 0; } else { return 1; }"
       "}",
```
Same below.

lenticularis39 replied:
I don't think clang-format is enforcing this, I just let it auto-format the file and it did it like that.

viktormalik reviewed (commented):
Looks good! I mostly have some minor suggestions.

lenticularis39 commented on src/ast/passes/return_path_analyser.cpp:
```diff
@@ -0,0 +1,88 @@
+
+bool ReturnPathAnalyser::visit(Program &prog)
+{
+  for (Subprog *subprog : *prog.functions)
```
Not only this file, it's a common ugliness of my code :D

lenticularis39 commented on src/ast/ast.cpp:
```diff
@@ -204,7 +221,14 @@ Probe::~Probe()
+    for (SubprogArg *a : *args)
+      delete a;
+  delete args;
   if (stmts)
     for (Statement *s : *stmts)
       delete s;
```
Ah I see, there is no double free, since I set `stmts` to `nullptr` so I didn't notice. Nice catch!

danobi commented:
Took a close look at ast, parser, lexer, return_path_analyser. They look really nice. Will try to make another pass tmr morning over semantic analyser and codegen details.

danobi commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2160,7 +2197,21 @@ void SemanticAnalyser::visit(Jump &jump)
+        jump.return_value->accept(*this);
+      if (auto subprog = dynamic_cast<Subprog *>(scope_))
+      {
+        if ((subprog->return_type.IsVoidTy() !=
+             (jump.return_value == nullptr)) ||
```
Nit: easier to read if `is_void && (return_value != nullptr)` IMO

danobi reviewed (commented):
semantic analysis lgtm; one small nit

danobi commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2582,6 +2587,50 @@ void CodegenLLVM::generateProbe(Probe &probe,
+  BasicBlock *entry = BasicBlock::Create(module_->getContext(), "entry", func);
+  b_.SetInsertPoint(entry);
+
+  variables_.clear();
+  ctx_ = nullptr;
```
Would it make sense to hoist this out to a helper function? I think `generateProbe()` shares the `variables_.clear()` bit. It would help self-document the intent and make future extensions less error prone

lenticularis39 replied:
Parts of it are shared, but parts of it are also different, subprogs process arguments, probes do not, probe do additional setup subprogs don't.

danobi commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2582,6 +2587,50 @@ void CodegenLLVM::generateProbe(Probe &probe,
+  inside_subprog_ = true;
+
+  int arg_index = 0;
+  for (SubprogArg *arg : *subprog.args)
+  {
+    auto alloca = b_.CreateAllocaBPF(b_.GetType(arg->type), arg->name());
+    b_.CreateStore(func->getArg(arg_index), alloca);
+    variables_.insert({ arg->name(), alloca });
+    ++arg_index;
+  }
```
Is it necessary to put arguments on stack? I know `variables_` value is an allocainst. But I'm not sure it's strictly necessary (unless I'm mistaken which I could easily be). AFAIU should be OK to keep variables in register.

lenticularis39 commented on src/ast/passes/codegen_llvm.h:
```diff
@@ -166,7 +168,7 @@ class CodegenLLVM : public Visitor
   // Create return instruction
   //
-  // If null, return value will depend on current attach point
+  // If null, return value will depend on current attach point (void in subprog)
```
Subprogs are allowed to return a value, the comment only refers to `createRet()` being called without an argument/with null argument.

danobi replied:
Ah yeah. Misread the comment

lenticularis39 commented:
Just fixed a small readability issue (moving check for void return value to `ReturnPathAnalyser::visit(Subprog &subprog)`).

lenticularis39 commented on src/ast/passes/return_path_analyser.cpp:
```diff
@@ -0,0 +1,100 @@
+{
+  for (Subprog *subprog : *prog.functions)
+  {
+    if (subprog->return_type.type != Type::voidtype && !visit(*subprog))
```
Fixed.

viktormalik reviewed (approved):
Thanks!

@ajor wants to review this, too, so let's wait for that before merging.

ajor commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -377,7 +391,8 @@ void SemanticAnalyser::visit(Builtin &builtin)
   }
   else if (builtin.ident == "func") {
-    for (auto &attach_point : *probe_->attach_points)
+    auto probe = dynamic_cast<Probe *>(scope_);
+    for (auto &attach_point : *probe->attach_points)
```
I'm guessing these bits of code are the reason for disallowing certain builtins in `check_available_in_subprog()`?

I think it'd be safer if we included the checks closer to the point where they're actually needed (here) - I can imagine in the future we'll forget about this implicit link, modify the list in `check_available_in_subprog` and end up with a SEGV from this dynamic_cast returning a null pointer.

i.e. something like:
```suggestion
    auto *probe = dynamic_cast<Probe *>(scope_);
    if (!probe)
    {
      LOG(ERROR, builtin.loc, err_) << builtin.ident << " cannot be used in functions";
      return;
    }
    for (auto &attach_point : *probe->attach_points)
```

ajor reviewed (changes requested):
This is really looking pretty good now, thanks!

I just found some bugs that we should fix before merging.

Crash using disallowed builtins in functions:
```
$ sudo ./bpftrace -e 'fn foo(): int64 { return func; }' 
Segmentation fault
```

Crash writing to a map in a function:
```
$ sudo ./bpftrace -e 'fn bar(): void { @x = 1; }'
bpftrace: /home/ajor/src/bpftrace/src/ast/irbuilderbpf.cpp:444: void bpftrace::ast::IRBuilderBPF::CreateMapUpdateElem(llvm::Value*, bpftrace::ast::Map&, llvm::Value*, llvm::Value*, const bpftrace::location&): Assertion `ctx && ctx->getType() == getInt8PtrTy()' failed.
Aborted
```

Crash using map's value in a function:
```
$ sudo ./bpftrace -e 'fn bar(): uint64 { $res = @x + 1; return $res; } BEGIN { @x = 1; }'
bpftrace: /home/ajor/src/bpftrace/src/ast/irbuilderbpf.cpp:379: llvm::Value* bpftrace::ast::IRBuilderBPF::CreateMapLookupElem(llvm::Value*, bpftrace::ast::Map&, llvm::Value*, const bpftrace::location&): Assertion `ctx && ctx->getType() == getInt8PtrTy()' failed.
Aborted
```

Poor error message reading a map from a function:
```
$ sudo ./bpftrace -e 'fn bar(): int64 { return @x; } BEGIN { @x = 1; }'
stdin:1:20-29: ERROR: Function bar is of type int64, cannot return none
fn bar(): int64 { return @x; } BEGIN { @x = 1; }
                   ~~~~~~~~~
```

lenticularis39 commented on src/ast/passes/printer.cpp:
```diff
@@ -377,6 +383,32 @@ void Printer::visit(Probe &probe)
+  out_ << indent << subprog.name() << ": " << subprog.return_type;
+
+  out_ << "(";
+  if (subprog.args->size() > 0)
+  {
```
Also I noticed the function outputs the old syntax of arguments:
```
# build/src/bpftrace -de 'fn test($x: int64): void {  } BEGIN {}'
[DEBUG] libbpf: loading kernel BTF '/sys/kernel/btf/vmlinux': 0
[DEBUG] libbpf: loading kernel BTF '/sys/kernel/btf/vmlinux': 0

AST after: parser
-------------------
Program
 test: void(int64 $x)
 BEGIN
```

lenticularis39 commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2582,6 +2587,50 @@ void CodegenLLVM::generateProbe(Probe &probe,
+  if (subprog.return_type.IsVoidTy())
+    createRet();
+
+  FunctionPassManager fpm;
+  FunctionAnalysisManager fam;
+  llvm::PassBuilder pb;
+  pb.registerFunctionAnalyses(fam);
+  fpm.addPass(UnreachableBlockElimPass());
+  fpm.run(*func, fam);
```
Unlike probes which have a default return value, subprograms might leave unreachable blocks that are not finished with a return instruction, these have to be removed for the IR to be correct.

ajor replied:
Just been thinking about this some more and I realise that I don't actually understand. How can a subprogram generate an unreachable block without a return instruction?

lenticularis39 commented:
Fixed maps and casts in subprogs. Also added more tests to cover this functionality.

Subprog argument implementation in codegen was changed to include ctx as the first argument, which is needed for perf-event-based builtin functions (like print) to work properly. This was tested at runtime with the in-progress function call PR.

With this all feedback should be resolved.

danobi commented:
Are there any outstanding work items left? Or are we missing a final round of reviews?

Side note, this behavior is a bit unexpected:

```
$ sudo ./build/src/bpftrace -e 'fn foo($a: int32): int32 { return $a + 4; }'
stdin:1:30-43: ERROR: Function foo is of type int32, cannot return int64
fn foo($a: int32): int32 { return $a + 4; }

```

I suspect this is more related to how we represent all ints as int64 internally. Not a show stopper -- IMO fixing this relaxes some rules so this is ok to do later.

danobi commented:
Also another bit of fix-later feedback:

```
sudo ./build/src/bpftrace -e 'fn foo(): void { return; }'
No probes to attach

$ sudo ./build/src/bpftrace -e 'fn foo() { return; }'
stdin:1:10-11: ERROR: syntax error, unexpected {, expecting :
fn foo() { return; }

```

Going with bpftrace's philosophy of brevity, perhaps good to allow `void` return types to be omitted. But I think this is debatable

lenticularis39 commented:
> Are there any outstanding work items left? Or are we missing a final round of reviews?
> 
No, just waiting for reviews.

> Side note, this behavior is a bit unexpected:
> 
> ```
> $ sudo ./build/src/bpftrace -e 'fn foo($a: int32): int32 { return $a + 4; }'
> stdin:1:30-43: ERROR: Function foo is of type int32, cannot return int64
> fn foo($a: int32): int32 { return $a + 4; }
>                              ~~~~~~~~~~~~~
> ```
> 
> I suspect this is more related to how we represent all ints as int64 internally. Not a show stopper -- IMO fixing this relaxes some rules so this is ok to do later.
> 
> Also the error underline seems to be off.

I know about this, it's because types were not explicitly given anywhere previously (besides casts). Not sure why the underlining is off.

ajor commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2372,12 +2376,60 @@ void CodegenLLVM::generateProbe(Probe &probe,
     b_.CreateRet(value);
     return;
+  } else if (inside_subprog_) {
+    b_.CreateRetVoid();
```
I'm assuming this is just a temporary thing until your next PR?

lenticularis39 replied:
This is because unlike with probes, there is not any default return value for subprogs. createRet() without argument is only used to end unreachable code that is then removed in subprogs.

ajor reviewed (approved):
Sorry for taking so long to review this. Really appreciate the work you've put it!

The code's looking good to me - the comments are just a bunch of minor things that can be fixed later. Let's just get this one through without more delay!

The only issue I found was that we're allowing multiple functions to be defined with the same name. Again, this can be fixed later:
```
fn aaa(): void { print(1); }
fn aaa(): void { print(2); }
```

lenticularis39 commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2105,8 +2182,13 @@ void SemanticAnalyser::visit(Cast &cast)
+      ProbeType type = single_provider_type(probe);
+      cast.type.SetAS(find_addrspace(type));
+    } else {
+      // Assume kernel space for data in subprogs
```
Yes, the issue is this probably requires a new language feature, since in the context of a subprog, there is no way to tell if you are accessing userspace or kernel space data based on pointers alone.

lenticularis39 commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2372,12 +2376,60 @@ void CodegenLLVM::generateProbe(Probe &probe,
+                 [this](SubprogArg *arg) { return b_.GetType(arg->type); });
+  FunctionType *func_type = FunctionType::get(b_.GetType(subprog.return_type),
+                                              arg_types,
+                                              0);
```
Yeah I must have copied this from somewhere :D

lenticularis39 commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -879,7 +878,8 @@ void CodegenLLVM::visit(Call &call)
   } else if (call.func == "printf") {
     // We overload printf call for iterator probe's seq_printf helper.
-    if (probetype(current_attach_point_->provider) == ProbeType::iter) {
+    if (!inside_subprog_ &&
```
Currently subprogs do not track if they are called from an iterator probe so printf from subprog from iterator probe is treated like outside iterator probe. This should be probably fixed eventually.

lenticularis39 commented:
The PR should be now rebased and ready for merge.

ajor commented:
Thanks @lenticularis39!

ajor commented on tests/semantic_analyser.cpp:
Added in #3062
